core/src/aggregator.rs

```diff
@@ -81,13 +86,100 @@ impl Aggregator {
             .unwrap();
 
         Ok(Aggregator {
+            db,
             config,
             nofn_xonly_pk,
             verifier_clients,
             operator_clients,
         })
     }
 
+    /// Creates a stream of nonces from verifiers.
+    /// This will automatically get's the first response from the verifiers.
+    pub async fn create_nonce_streams(
+        &self,
+        num_nonces: u32,
+    ) -> Result<
+        (
+            Vec<clementine::NonceGenFirstResponse>,
+            Vec<BoxStream<Result<MuSigPubNonce, BridgeError>>>,
+        ),
+        BridgeError,
+    > {
+        // generate nonces from all verifiers
+        let mut nonce_streams = try_join_all(self.verifier_clients.iter().map(|client| {
+            // Clone each client to avoid mutable borrow.
+            // https://github.com/hyperium/tonic/issues/33#issuecomment-538150828
+            let mut client = client.clone();
+
+            async move {
+                let response_stream = client
+                    .nonce_gen(tonic::Request::new(clementine::NonceGenRequest {
+                        num_nonces,
+                    }))
+                    .await?;
+
+                Ok::<_, Status>(response_stream.into_inner())
+            }
+        }))
+        .await?;
+
+        // Get the first responses from each stream
+        let first_responses: Vec<clementine::NonceGenFirstResponse> =
+            try_join_all(nonce_streams.iter_mut().map(|s| async {
+                let nonce_gen_first_response = s.message().await?;
+                let nonce_gen_first_response = nonce_gen_first_response
+                    .ok_or(BridgeError::Error("NonceGen returns nothing".to_string()))?
+                    .response
+                    .ok_or(BridgeError::Error(
+                        "NonceGen response field is empty".to_string(),
+                    ))?;
+                // this response is an enum, so we need to match on it
+                match nonce_gen_first_response {
+                    clementine::nonce_gen_response::Response::FirstResponse(
+                        nonce_gen_first_response,
+                    ) => Ok(nonce_gen_first_response),
+                    _ => Err(BridgeError::Error(
+                        "NonceGen response is not FirstResponse".to_string(),
+                    )),
+                }
+            }))
+            .await?;
+
+        // Transform each stream's output using map
+        fn extract_pub_nonce(
```

Comment:
Fixed in https://github.com/chainwayxyz/clementine/pull/361/commits/1bb5bd1bb4cb959bcff327b56813a2f5ad2b5a12. No unit tests are written for it, though. Because it is a simple function that is already getting tested by the only caller.